libraries/ms-common/src/stm32f0xx/can_hw.c

```diff
@@ -0,0 +1,172 @@
+#include "can_hw.h"
+#include "stm32f0xx.h"
+
+#define CAN_HW_PRESCALER 12
+
+static volatile CANHwConfig *s_can = NULL;
+
+StatusCode can_hw_init(CANHwConfig *can_hw, uint16_t bus_speed, bool loopback) {
+  CAN_DeInit(CAN);
+  memset(can_hw, 0, sizeof(*can_hw));
+
+  can_hw->base = CAN;
+  can_hw->bus_speed = bus_speed;
+  can_hw->loopback = loopback;
+
+  s_can = can_hw;
+
+  NVIC_InitTypeDef nvic_cfg = {
+    .NVIC_IRQChannel = CEC_CAN_IRQn,
+    .NVIC_IRQChannelPriority = 0,
+    .NVIC_IRQChannelCmd = ENABLE
+  };
+  NVIC_Init(&nvic_cfg);
+
+  RCC_APB1PeriphClockCmd(RCC_APB1Periph_CAN, ENABLE);
+
+  RCC_ClocksTypeDef clocks;
+  RCC_GetClocksFreq(&clocks);
+  // time quanta = (APB1 / Prescaler) / baudrate, -1 for start bit
+  uint16_t tq = clocks.PCLK_Frequency / 1000 / CAN_HW_PRESCALER / can_hw->bus_speed - 1;
+  uint16_t bs1 = tq * 7 / 8; // 87.5% sample point
+
+  CAN_DeInit(can_hw->base);
```

Comment:
What's the difference between this line here and when you call `CAN_DeInit(CAN);` at the start of the function?